proposal: CephFS snapshots as shallow RO volumes

gman0 opened the pull request:
This PR adds a proposal document for 'CephFS snapshots as shallow read-only volumes' feature.

Related issue: https://github.com/ceph/ceph-csi/issues/2142

gman0 committed:
doc: add proposal doc for CephFS snapshots as shallow RO volumes

This patch adds a proposal document for "CephFS snapshots
as shallow RO volumes".

Updates: #2142
Signed-off-by: Robert Vasek <robert.vasek@cern.ch>

nixpanic commented:
This looks well thought out :+1: 

Obviously the reference counting is the most important factor for this. It would be the main part where the complexity needs to be handled. I am not sure the additional complexity is worth the benefits.

Would this be something that will be used a lot in certain type of workloads? The use-cases in the document are very clear, but I would like to know more about the actual applications and increase of performance (why is it needed, or is it just nice to have?) for these.

gman0 commented:
@nixpanic I'm discussing this proposal with Kubernetes Data Protection WG still. I'll try to give more concrete examples soon.

gman0 commented:
And just to make it clear, I wouldn't be trying to solve the "multi-master race". At least not as a part of this proposal anyway. There are a few other places in the driver where this is a problem too, no? Or at least I remember reading something along these lines... It would be nice if we could solve it for all these cases at once.

humblec commented:
@gman0  The  proposal looks good to me :+1:   the shallow RO volume mount for an export make sense to transfer a backup to external source.  One caveat here is fitting this workflow on  data protection w efforts in general which you already taken into account and figuring out .

gman0 commented:
@nixpanic after checking with Data Protection WG, @xing-yang, this proposal should fit in nicely with the planned Volume backup feature - which itself is a part of bigger Application backup feature. It's desirable for a CSI driver to be able to create volumes from snapshots efficiently. All the heavy I/O would be done only when copying the data to the backup/archive storage. This is as opposed to having to create a clone of the snapshot just so that it can be read from when transferring to backup storage. Depending on the amount of data and current load on the Ceph cluster, the increase in performance may be significant.

Right now this is usable with Velero/Restic. We're planning to use this for regular volume/app backups for our users. Later, I'd like to propose to Velero folks to include "volume from snapshot" as an intermediate step when performing volume backups using their [CSI plugin](https://github.com/vmware-tanzu/velero-plugin-for-csi).

gman0 commented:
I've squashed commits for fixing markdown linter errors and added a line saying that solving the race condition for when there are multiple controller plugins running concurrently is out of scope for this proposal. Commit lint is still complaining though, something about subject being empty. I'm not sure what's up with that.

gman0 commented:
@nixpanic does this seem reasonable to you?

gman0 commented:
Is there anything else needed for this PR? @Madhu-1 @ShyamsundarR @humblec PTAL

Madhu-1 commented:
commitlint is failing as the commit description is not present. https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages

Madhu-1 commented:
Busy with some work, I will revisit this soon. Thanks, @gman0

gman0 commented:
@Madhu-1 thanks. I can squash the commits that fix the linter errors with the original commit later when we're done with reviews. That should satisfy commitlint too I hope.

Madhu-1 commented on docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md:
```diff
@@ -0,0 +1,274 @@
+protected by mutex locks and is easy to implement. `cephfs.ControllerServer`
+already offers `VolumeLocks`, `SnapshotLocks` and `OperationLocks`.
+
+Mutexes won't cover situations where multiple masters may be active though
+(during/after leader election), and modifying _refcount_ would still race in
+this case. Misbehaving master that lost leadership would still be able to
+access resources on the Ceph cluster. Racing on _refcount_ could then cause
+leaking the _refcount_ object itself and/or the subvolume snapshot, or the
+provisioned shallow volume could point to a subvolume snapshot that's already
+been deleted. This is not easily fixable without the individual master replicas
+being aware of the leader election process and is out-of-scope for this
+proposal.
+
```
Currently, also this problem exists. but this could be a bigger problem as we don't have any lease coordination with the sidecar we are running. as we allow multiple replicas in provisioner deployment. in one pod the external-provisioner might be a leader and in the other pod, the external-snapshotter is acting as a leader?

gman0 replied:
Yeah I was thinking about this scenario too. I see that in another commet you already suggested setting Deployment replicas to 1. I'm wondering whether it shouldn't be the default anyway though, and whether it's enough to protect from similar races in general. With Deployments r=1 there still could be more than one replica running concurrently (albeit the others would be in Terminating state), so there's still a chance for a race to occur no? StatefulSets provide the uniqueness guarantee which would take care of this problem altogether.

Whenever ceph-csi needs a mutex to protect Ceph resources from concurrent access, it's usually a good indicator some locking / fencing between controller replicas is also necessary, should a leader switch occur. What leader election provides in sidecars is simply not enough:

https://github.com/kubernetes/client-go/blob/release-14.0/tools/leaderelection/leaderelection.go#L17-L20

How much availability would ceph-csi lose if it were to simply ditch leader election and use a single-replica StatefulSets instead? Please correct me if I'm wrong, but isn't the behavior for resolving failure with and without leader election very similar? 
* With leader election: leader controller plugin crashes / becomes unresponsive, another replica is selected as the new leader.
* Without leader election: controller plugin crashes / becomes unresponsive (signaled by a liveness probe), Kubernetes takes care of recreating the Pod.

In both of those cases some unavailability is present whether it is due to timing out and finding a new leader, or creating a new Pod. Except that multiple replicas + leader election run the risk of race conditions. Maybe we can discuss this more in a separate ticket.

Madhu-1 replied:
Yes it's a bit tricky and needs a lot of testing. we moved from statefulset to deployment due to the downtime issues statefulset. and deployment with replica:2/3 got added to have the HA. but with this one, we have a leader election problem. @ShyamsundarR opened an issue for this one https://github.com/ceph/ceph-csi/issues/742 I think we need to check with the CSI team as this can be a problem? but if CSI is stateless this does not matter but for us, it matters as things can go wrong with the counter.

gman0 replied:
Thanks for the pointer, @Madhu-1 ! I suppose this is the problem you've mentioned https://github.com/ceph/ceph-csi/pull/414 ?

Madhu-1 replied:
Yes, correct that's the one.

gman0 replied:
With atomic reference counting / tracking, I think we no longer face the issues mentioned in this thread and in the paragraph regarding race conditions. It no longer matters how many readers/writers modify the references. It also means the requirement for single replica deployments can be lifted.

gman0 commented:
Any updates on reviews on this PR please?

ShyamsundarR commented on docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md:
```diff
@@ -0,0 +1,274 @@
+  that shallow volume still exists?_
+
+  We need to make sure this doesn't happen and some book-keeping
+  is necessary. Ideally we could employ some kind of reference counting.
```
I read till here, and this is a useful proposal. I did have similar thoughts earlier but did not take it forward, so thanks @gman0 .

I will review the rest next week. Need to look at this from a CephFS subvolume perspective on snapshot locations and snap path exposed/staged for workloads (just a note to self).

gman0 replied:
@ShyamsundarR cool, thanks!

Madhu-1 commented on docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md:
```diff
@@ -0,0 +1,274 @@
+  We need to make sure this doesn't happen and some book-keeping
+  is necessary. Ideally we could employ some kind of reference counting.
+
+#### Reference counting for shallow volumes
```
Do we need to consider PVC-PVC clone case from PVC creating as shallow volume? to allow disallow the PVC-PVC clone operation?

gman0 replied:
Good catch!

* request to create **shallow volume** with **shallow volume** as its source: basically a new reference to the parent snapshot,
* request to create **regular volume** with **shallow volume** as its source: equivalent for a request to create a regular volume with snapshot as its source,
* request to create **shallow volume** with **regular volume** as its source: this doesn't make sense and should exit with an error.

gman0 replied:
https://github.com/ceph/ceph-csi/pull/2148/commits/4db796a7decf17050bfd66fb3ed9c54ebe0acb68

Madhu-1 commented on docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md:
```diff
@@ -0,0 +1,274 @@
+* _What should be the requested size when creating the volume (specified e.g.
+  in PVC)?_
+
+  This one is tricky. CSI spec allows for
```
Yes even though there is no consumption we need to return the requested size in response to make sure no Error happens at the kubernetes level.

Madhu-1 commented on docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md:
```diff
@@ -0,0 +1,274 @@
+Snapshotting read-only volumes doesn't make sense in general, and should
+be rejected.
+
+### `ControllerExpandVolume`
+
+Same thing as above. Expanding read-only volumes doesn't make sense in general,
+and should be rejected.
```
from the kubernetes point of view, it's a new PVC. consider adding CreateVolume from snapshot and CreateVolume from volume case here as well.

Madhu-1 commented on docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md:
```diff
@@ -0,0 +1,274 @@
+provisioned shallow volume could point to a subvolume snapshot that's already
+been deleted. This is not easily fixable without the individual master replicas
+being aware of the leader election process and is out-of-scope for this
+proposal.
```
Add a note in the document when we implement this one. To use this feature and to avoid issues run Deployment with replica `1`

gman0 replied:
https://github.com/ceph/ceph-csi/pull/2148/commits/b02a288ecad4694cfaf87d770f828c7c9191584a

Madhu-1 commented on docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md:
```diff
@@ -0,0 +1,274 @@
+  [`NodeStageVolume`, `NodeUnstageVolume` section](#NodeStageVolume-NodeUnstageVolume),
+  snapshots cannot be mounted directly. How do we pass in path to the parent
+  subvolume?
+  * a) Path to the snapshot is passed in via `subvolumePath` / `rootPath`,
```
to keep things simple let's have an extra path for snapshot also and this will be used only in the NodePublish right?

gman0 replied:
:+1: yes

gman0 replied:
When do we look up the path of the snapshot itself?

The dirname has following format: `_<SNAP NAME>_<INODE NUMBER>`. We don't know what it is the directory called unless we actually iterate through `.snap` and find the correct path, i.e. the directory whose name contains `<SNAP NAME>`.

https://github.com/ceph/ceph/blob/a17ebc0406ac5592911fa7f596464c9c848a9c1e/src/mds/snap.cc#L81-L91

When?
* a) In `CreateVolume`. We'd need to mount the volume, look up the snapshot dir, and then unmount the volume. Dirname would be then stored in the response, in `Volume.volume_context["snapshotDirname"]`.
   * Pros: lookup is done only once per volume.
   * Cons: pre-provisioned volumes would have to do the lookup themselves in order for them to populate `Volume.volume_context["snapshotDirname"]`.
* b) In `NodeStageVolume`. `Volume.volume_context["snapshotName"]` must be populated. In this case we'd also need to mount the volume, but we'd need to do that either way. The only extra bit of work is doing the lookup once per PVC per node.
   * Pros: dynamically provisioned and pre-provisioned volumes are treated the same way.
   * Cons: one extra directory lookup once per PVC per node. Probably negligible performance penalty.
* c) Combination of a) and b). `CreateVolume` would populate `Volume.volume_context["snapshotDirname"]`. If it's not present in volume context during `NodeStageVolume` (in case of pre-provisioned volumes), but `snapshotName` is, it would do the lookup then.

Madhu-1 replied:
If the subvolume path is `csi-vol-1d17e149-f522-11eb-938d-866c88286c89/e50a2a89-a4fb-4fca-b590-8e8476e8a1a6/` and the snapshot path is present at `csi-vol-1d17e149-f522-11eb-938d-866c88286c89/.snap/csi-snap-f0a7e7be-f524-11eb-938d-866c88286c89` with this we can mount the volume at the stagingPath? what do you think?

Madhu-1 replied:
cc @kotreshhr

gman0 replied:
Ah right, snapshot naming is different in parent dir. Thanks.
```
# pwd
.../volumes/csi/csi-vol-9dba52f1-f54b-11eb-b0aa-0242ac110006
# ls .snap
csi-snap-b7063583-f54b-11eb-b0aa-0242ac110006
```

The problem is when provisioning is managed by an external service (e.g. OpenStack Manila). In such environments the cephx clients have access restricted to the subvolume path only. `.snap` in parent is inaccessible, so they'd need to go to `subvol/.snap`:
```
# pwd
.../volumes/csi/csi-vol-9dba52f1-f54b-11eb-b0aa-0242ac110006/e6bb7726-71f3-4d32-940c-c34ea08ed383
# ls .snap
_csi-snap-b7063583-f54b-11eb-b0aa-0242ac110006_2199023257056
```
And here you can see the `_<SNAP NAME>_<INODE NUMBER>` name format I mentioned in my comment above. Since we don't know the ino num beforehand, *somebody* needs to do a name lookup.

gman0 replied:
So for volumes provisioned by cephfs-csi, dirname is trivial to figure out.

Who should do snap dirname lookup for pre-provisioned volumes / volumes managed externally?
* a) The consumer. cephfs-csi would take dirname via volume context. This means that the consumer needs to mount the volume, perform the lookup, unmount it, and then create a pre-provisioned shallow volume with dirname specified.
* b) cephfs-csi. Since cephfs-csi already needs to mount the whole volume, it could easily do the dirname lookup itself. In this case users would only need to supply the snapshot name.

Which do you think makes more sense? b) is obviously a lot more user-friendly -- at the price of a bit more complexity.

Madhu-1 replied:
for pre-provisioned volume, i think we can go for `option a` to keep it simple and as the user is manually creating it admin can populate the volumeContext with required snap details so cephcsi will mount it

Madhu-1 reviewed (commented):
overall LGTM. left few comments

humblec commented on docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md:
```diff
@@ -0,0 +1,274 @@
+  restoring data to a writable volume more selectively instead of restoring
+  the whole snapshot.
+* **Volume backup:** users can't backup a live volume, they first need
+  to snapshot it. Once a snapshot is take, it still can't be backed-up,
```
is taken?

gman0 replied:
Oops, that's a typo.

gman0 replied:
https://github.com/ceph/ceph-csi/pull/2148/commits/a98a5d30d69c2b25c80edfb413eeb5bd79c7af1a

humblec commented on docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md:
```diff
@@ -0,0 +1,274 @@
+
+Key points:
+
+* Volume source is a snapshot, volume access mode is `*_READER_ONLY`.
```
This `*_READER ONLY` is bit complex or an area at kubernetes where things are vague or incomplete. However there are recent improvements on this , but yet to be completed. so I am not sure  is it a good idea to rely on READER ONLY access mode. In other side, considering this is a design proposal may be we are good and later while coming into implementation we can check thoroughly.

gman0 replied:
I didn't want to pollute storage class parameters any more than necessary. I don't have a test implementation of this yet so I'm not sure 100% either about this. If you're ok with this for now, we can see during implementation, like you've suggested.

humblec replied:
The problem is that, eventhough the request is for readonly, the node mount is "read write", are we fine with that?

gman0 replied:
We don't care what users specify as access mode or if it's read-write. It will always be exposed as read-only to workloads. We can make stricter validation and refuse mounting volumes that want to be RW and have snapshot ref in their volume context parameters.

humblec commented on docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md:
```diff
@@ -0,0 +1,274 @@
+* Volume source is a snapshot, volume access mode is `*_READER_ONLY`.
+* No actual new subvolumes are created in CephFS.
+* The resulting volume is a reference to the source subvolume snapshot.
+  This reference would be demarked in `Volume.volume_context` map.
```
can you please explain some more on , what is meant here by `demarked` ?

gman0 replied:
I'll use a different word and clarify what the reference looks like.

gman0 replied:
https://github.com/ceph/ceph-csi/pull/2148/commits/b68dd5ebbfee3c3fe75a5341ce69b2decd3b0217

gman0 commented:
I think I've addressed the comments we had so far, PTAL.

Madhu-1 reviewed (commented):
I don't have any major comments except what we already discussed. good to go for implementation we can update the design(minor changes) if required when we start the implementations

@gman0 please squash the commits and make the CI happy :)

gman0 commented:
@Madhu-1 thanks! Commits squashed and all CI jobs have passed.

Madhu-1 commented:
@humblec @nixpanic PTAL

haslersn commented:
> Later, I'd like to propose to Velero folks to include "volume from snapshot" as an intermediate step when performing volume backups using their [CSI plugin](https://github.com/vmware-tanzu/velero-plugin-for-csi).

@gman0 I think that proposal exists: https://github.com/vmware-tanzu/velero/issues/3544

gman0 commented:
@haslersn great, I wasn't aware of this. Thank you for letting us know!

gman0 commented:
@nixpanic @humblec @ShyamsundarR any updates on this proposal? I guess we can always update it later if we need to.

humblec commented:
> @nixpanic @humblec @ShyamsundarR any updates on this proposal? I guess we can always update it later if we need to.

@gman0  Apologies. Will take a final look at this tomorrow without fail..

gman0 commented:
@humblec thanks! I'm now working on adding a couple of RADOS bindings to go-ceph, for atomic write ops, so that at least the reference counter itself is race-free.

humblec commented on docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md:
```diff
@@ -0,0 +1,297 @@
+* Volume source is a snapshot, volume access mode is `*_READER_ONLY`.
+* No actual new subvolumes are created in CephFS.
+* The resulting volume is a reference to the source subvolume snapshot.
+  This reference would be stored in `Volume.volume_context` map. In order
```
this is specific to node stage request , isnt it ?

gman0 replied:
Well any call that has `volume_context` in request, and when we need to distinguish between volume kinds. But yes, also when staging the volume.

humblec commented on docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md:
```diff
@@ -0,0 +1,297 @@
+  This reference would be stored in `Volume.volume_context` map. In order
+  to reference a snapshot, we need subvol name and snapshot name.
+* Mounting such volume means mounting the respective CephFS subvolume
+  and exposing the snapshot to workloads.
```
just wondering what if the snapshot get deleted while the subvolume is in use or mounted. because kubelet or ocp does not know at this stage whether the inside snap is getting read by someone in this scenario. did I miss anything here ?

gman0 replied:
This is mentioned in [this](https://github.com/ceph/ceph-csi/blob/3edc507537d7ae12933ec750cb5efb6d14397c38/docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md#controller-operations) paragraph:

> When a shallow volume is created, what would happen if parent subvolume of the snapshot is removed while the shallow volume still exists?
>
> This shouldn't be a problem already. The parent volume has either `snapshot-retention` subvol feature in which case its snapshots remain available, or if it doesn't have that feature, it will fail to be deleted because it still has snapshots associated to it.

gman0 replied:
So either the snapshot will be kept, or the subvolume deletion will be refused by Ceph. In both cases we should be safe.

humblec commented on docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md:
```diff
@@ -0,0 +1,297 @@
+the number of references the backing subvolume snapshot has. It would be
+initialized to one, where the CSI snapshot itself is the first reference to the
+backing snapshot. Each subsequent shallow volume creation would increment the
+reference counter, and each shallow volume deletion would decrement it.
```
if we reference count only on `creation and deletion`, does that handle a situation of "multiple readers" of the same swallow volumes and thus consider deletion when all the references are gone?

gman0 replied:
I hope I understood the question correctly, but yes, it doesn't matter how many consumers of the shallow volume there are.

humblec replied:
ok.. thanks for confirming

gman0 commented on docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md:
```diff
@@ -0,0 +1,297 @@
+snapshot be requested for deletion.
+
+When creating a volume snapshot, a new RADOS object would be created for that
+snapshot. It would contain a single signed integer value that would describe
```
Integer-based reference counter would break idempotency.

We have no way of distinguishing repeated calls on `CreateVolume`. We'd have duplicate counts. The same problem arises with *Delete ops.

Obviously we still need to track and count references, so this part of the proposal is ok I think, I should have left out the implementation details out though. We may need to store volume IDs (in an omap perhaps) and derive the number of references from that.

humblec replied:
Correct, integer based reference counters are error prone. if the plan is to put it on volumeIDs or other unique entities, we are good @gman0

gman0 commented:
I'll squash and rebase again. @Madhu-1 @humblec can you please take a look on the latest changes? I've reworked the "reference counter" into "reference tracker" - see the comment here https://github.com/ceph/ceph-csi/pull/2148#discussion_r726134474 and here https://github.com/ceph/ceph-csi/pull/2148#discussion_r728029855 .

gman0 commented:
Just FYI, following with the latest findings, I've created a draft implementation of the reference tracker + accompanying program for testing -- albeit in C just for now. Once there's proper support from go-ceph, porting this to Go should be trivial.

https://github.com/gman0/rados-reference-tracker

This solves the problem with maintaining idempotency I mentioned in a comment above.

Madhu-1 commented on docs/design/proposals/cephfs-snapshot-shallow-ro-vol.md:
```diff
@@ -0,0 +1,272 @@
+
+#### Reference counting for shallow volumes
+
+As mentioned above, this is to protect shallow volumes, should their source
+snapshot be requested for deletion.
+
+When creating a volume snapshot, a reference tracker (RT), represented by a
+RADOS object, would be created for that snapshot. It would store information
+required to track the references for the backing subvolume snapshot. Upon a
+`CreateSnapshot` call, the reference tracker (RT) would be initialized with a
+single reference record, where the CSI snapshot itself is the first reference
+to the backing snapshot. Each subsequent shallow volume creation would add a
+new reference record to the RT object. Each shallow volume deletion would
+remove that reference from the RT object. Calling `DeleteSnapshot` would remove
+the reference record that was previously added in `CreateSnapshot`.
+
+The subvolume snapshot would be removed from the Ceph cluster only once the RT
+object holds no references. Note that this behavior would permit calling
+`DeleteSnapshot` even if it is still referenced by shallow volumes.
+
+* `DeleteSnapshot`:
+  * RT holds no references or the RT object doesn't exist:
+    delete the backing snapshot too.
+  * RT holds at least one reference: keep the backing snapshot.
+* `DeleteVolume`:
+  * RT holds no references: delete the backing snapshot too.
+  * RT holds at least one reference: keep the backing snapshot.
```
@gman0 why RT is consider during CreateSnapshot and DeleteSnapshot it should be considered only during CreateVolume and DeleteVolume isn't it? if it's a new CreateVolume request increment the RT count and for each DeleteVolume decrement the RT if RT is reached `0` remove the snapshot?

gman0 replied:
@Madhu-1 this decouples shallow volumes from their source CSI snapshots - it makes it possible to call `DeleteSnapshot` and succeeding.

Doing what you suggested shouldn't break other aspects of this proposal, so it is possible, but at the cost of `DeleteSnapshot` calls being stuck until there are no shallow volumes associated with the snapshot. This may be an unexpected behavior to users, who would expect `DeleteSnapshot` to finish right away. When users create thick-provisioned volumes from snapshots (i.e. the non-shallow ones), they are able to delete the source snapshot afterwards and they see it succeeding, while the volumes exist. I think this behavior should be mirrored for shallow volumes too, for consistency.

This is possible only when we count references in `CreateSnapshot` and `DeleteSnapshot` procedures too.

Madhu-1 replied:
I haven't looked How RT will be implemented but if it's a rados object, I was thinking of the below case.

* when RT key does not exists: During DeleteSnapshot, In this case, return delete the snapshot
* when RT exists but not `0`: In this case DeleteSnapshot checks for RT if there any RT return success as RT was added in the CreateVolume, the DeleteVolume will take care of cleaning up the snapshot once the RT becomes `0`

Madhu-1 replied:
this is not a blocker we can take care of it during the implementation.

gman0 replied:
@Madhu-1 ah I see. This is indeed more of an implementation detail of RT. We can discuss these when I'll push the ref-tracker PR.

Madhu-1 reviewed (commented):
changes LGTM. once the commits are squashed will approve it.

gman0 commented:
Thanks, @Madhu-1 ! Squashing changes.

gman0 commented:
@humblec @nixpanic any more comments on this PR? Can we get this merged?

humblec commented:
> @humblec @nixpanic any more comments on this PR? Can we get this merged?

With the explanations so far , we are good.. More or less, while implementing some adjustments are expected any way .  so lets get this in and start working on the implementation .. Thanks @gman0 ..

gman0 commented:
Thank you all for feedback! In order to implement this, it needs a couple of bindings added to go-ceph first -- for the reference tracker component. I'm currently working on them.

Rakshith-R commented:
/retest all

Rakshith-R commented:
> /retest all

Ci seems to be stuck https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/k8s-e2e-external-storage-1.21/detail/k8s-e2e-external-storage-1.21/1451/pipeline
